### ISSUE TITLE
Too many parts

### ISSUE THREAD

**Issue by larry-cdn77** [ClickHouse/ClickHouse#24102]
On my 15-node cluster (version 21.3.8) I host a number of replicated tables, 5 shards, 3 replicas each. Specifically, I have 3-5 million rows a second fed by a buffer table from multiple sources into a 'raw' MergeTree table. Off this hang a couple aggregating MergeTree tables with a local distributed table each for reading.

![2021-05-13 19 04 32](https://user-images.githubusercontent.com/72978202/118160099-29e7ae80-b41e-11eb-98f1-c2980324d832.png)

I have three aggregations across two of these chains, for a total of 30 ReplicatedMergeTree and 75 ReplicatedAggregatingMergeTree replicas running on the cluster

I had planned for more aggregations on top of the raw data but they are where I run into difficulty. Add one or two of these aggregation tables, and replication queues swell with fetch actions, 'too many parts' kicks in and data starts being lost. A plethora of other errors tag along, but replication queue and 'too many parts' seems to be at the root.

Remove an aggregation table or two and the cluster recovers itself fine

There is no apparent resource bottleneck. The underlying hardware hardly warms up with CPU, disk and network running low the entire time, and this is true for ClickHouse as well as ZooKeeper (a 5-node ZooKeeper instance, version 3.6.3).

Buffer table is set to flush every 1-4 seconds (16 layers and no size criteria). The raw table is partitioned by timestamp rounded to whole hour and configured to 10 hours of TTL. The few aggregations that I have running do simple counting and summing grouped by timestamp rounded to 5-second intervals, are partitioned to 5-minute intervals and have a TTL of 24 hours (similarly for the aggregations I am trying to add, I might have 1-minute grouping, 1-hour partitioning and 90-day TTL).

Quite a lot of effort has already gone into getting this to work, including these tweaked settings (some with specific reasons, some with blind faith):

```
┌─name─────────────────────────┬─value────────┐
│ max_block_size               │ 262144       │
│ max_insert_block_size        │ 4194304      │
│ use_uncompressed_cache       │ 0            │
│ background_pool_size         │ 32           │
│ background_move_pool_size    │ 32           │
│ background_fetches_pool_size │ 64           │
│ load_balancing               │ in_order     │
│ max_memory_usage             │ 100000000000 │
└──────────────────────────────┴──────────────┘
```

and:

```
┌─name────────────────────────────────────┬─value─────────┐
│ merge_max_block_size                    │ 65536         │
│ max_bytes_to_merge_at_max_space_in_pool │ 1099511627776 │
│ max_replicated_merges_in_queue          │ 64            │
│ max_replicated_mutations_in_queue       │ 32            │
│ max_replicated_merges_with_ttl_in_queue │ 4             │
│ parts_to_delay_insert                   │ 600           │
│ parts_to_throw_insert                   │ 1200          │
└─────────────────────────────────────────┴───────────────┘
```

A somewhat similar table architecture used to work on a small (5 nodes in a cross-segmented topology) cluster with an older release (20.8, I believe), with a dozen aggregations or so. I am in the process of recreating this older setup on the new big cluster with the main difference being an additional non-replicated table before the replicated raw table. Aggregations were done on the non-replicated table as well in this other setup.

Simply put, I have a setup that behaves as if there is a bottleneck (works until a certain number of tables) but no obvious bottleneck can be found; plus something similar used to work. Any suggestions would be greatly appreciated on whether this size cluster should comfortably sustain a number of 5x3-way replicated tables (2 raw and maybe a dozen aggregating) at 5 million rows a second. And if yes, why isn't it :-)

**Comment by den-crane** [comment]
check the output of this query https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-replication-queue

check this parameter `select * from system.settings where name = 'background_fetches_pool_size'　;`
If it's = 3, then increase it to 8 or upgrade to 21.3.9

**Comment by larry-cdn77** [comment]
Thank you for the prompt reply

I noticed 21.3.10 is an LTS release so will use that (unless there a specific reason not to and stick to 21.3.9)

Let me track the Altinity query when the problem occurs, condense into a meaningful summary and post back

**Comment by larry-cdn77** [comment]
The same setup described above is now running version 21.3.10

At a glance the replication pressure remained high and we continue to get 'too many parts'

The Altinity query shows that replication queue mostly consists of actions with the following characteristics:

- fetches
- postponed up to 100-200 times
- postpone reason '64 fetches already executing'
- occasionally reason is 'not executing because it is covered by part that is currently executing'
- no older than a minute or so
- never tried more than once or twice
- small number with exception 'no active replica has part'

**Comment by larry-cdn77** [comment]
I thought I would add draw total replication queue size to demonstrate how readily we observe the symptoms. 12:20 in the chart below is when I go from 5 aggregations to 3

![replicationQueue_3600sec](https://user-images.githubusercontent.com/72978202/118273722-94ebc080-b4c4-11eb-9a27-f0afb8e4739a.png)

**Comment by den-crane** [comment]
>Background fetches pool size has actually been set to 64 for a
> while now (I wonder if that could be too high a value for a 64-core machine)

There is no any sense to set it higher than 10.
Try to restart one of replica. It seems fetches just stuck. https://github.com/ClickHouse/ClickHouse/issues/23555

**Comment by larry-cdn77** [comment]
It seems that I had my buffer table set to flush just a bit more often than once a second, leading to excessive part count pressure

I had actually mistakenly assumed that zero in the buffer table min/max criteria means 'do not use' like so:

```Buffer('db', 'table', 16, 1, 5, 0, 0, 0, 0)```

What seems to [really happen](https://github.com/ClickHouse/ClickHouse/blob/2e3e56f3538f4b0ac9cfb58275ce2c0ff85754f9/src/Storages/StorageBuffer.cpp#L758) is that the criteria with zero in them always pass

I did not end up with thousands of inserts per second, more like two a second, but this was enough to cripple MergeTree, and replicated MergeTree in particular

I have re-entered all three buffering criteria, no longer see 'too many parts' and replication queue runs nice and low

Let me update this issue accordingly after running the cluster for some time

**Comment by larry-cdn77** [comment]
The good news is that making buffer table flush less frequently seemed to reduce the replication pressure and I was able to add aggregating tables until I had 14 (all triggered from the one MergeTree table as shown earlier)

After some hours I found a few nodes with inflated replication queues (and two of them doing 'too many parts')

To allow consuming this backlog I detached views of most aggregations and indeed, replication queues shrank down. Attached are two charts for illustration with the detach happening between 03:00 and 03:10.

![tooManyParts_3600sec](https://user-images.githubusercontent.com/72978202/118791346-eca87400-b896-11eb-8971-62bc3881137c.png)
![replicationQueue_3600sec](https://user-images.githubusercontent.com/72978202/118791364-f03bfb00-b896-11eb-94c0-31f647fa3d85.png)

A couple of nodes were left in a state where TTL merges on some of remaining live aggregating tables struggle to get their turn (postponed 100-200 thousand times with 'merges with TTL already executing'). Their number is slowly increasing so I wonder if these particular tables are somehow starved out of TTL merges.

Restarting those replicas did not help get their TTL merges through

My main question remains whether I am observing an expected limit of replicated aggregations (and if yes what they are). Any insights would be greatly appreciated.

**Comment by den-crane** [comment]
Not sure what TTL do you use? TTL delete? TTL move? TTL groupby?
In case of `TTL delete` you can use ttl_only_drop_parts.

> how many aggregating tables I can hang off one 1M rows/sec MergeTable

1M rows with 40 bytes each (4 columns) <>  1M rows with 4000 bytes each (900 columns).
Sometimes 2 Materialized Views are too many. Sometimes 40 Materialized Views are OK. All systems are different. I have a server with 128 CPUs and 2TB RAM and 10*2TB NVME.

**Comment by larry-cdn77** [comment]
I guess it's TTL delete, my aggregated tables are created like so:

```
CREATE TABLE db.table (
  t DateTime,
  s AggregateFunction(sum, UInt64)
)
ENGINE = ReplicatedAggregatingMergeTree('/clickhouse/tables/{shard}/t', '{replica}')
PARTITION BY toStartOfFiveMinute(t)
ORDER BY (t)
TTL t + INTERVAL 24 HOUR
```

Thank you for the suggestion, will try `ttl_only_drop_parts`

Taken your point about 'it depends', and it does indeed seem my cluster (64 cores/256 GB RAM) can sustain quite a few aggregations if configured properly

I will close the issue now and re-open if needed

**Comment by den-crane** [comment]
>PARTITION BY toStartOfFiveMinute(t)

Oh, wowwowowow. Mom, it's too scary.

**Comment by larry-cdn77** [comment]
Does it seem like too fine partitioning? I think it was the Altinity _query performance_ presentation where I picked up the partition count recommendation of 'hundreds not thousands' and therefore chose 24 hour TTL with 5-minute partitioning for this particular table for average of 288 live partitions. Maybe it makes sense to partition this aggregation table less or even not at all?
